### PR TITLE
ci: add rocksdb as a dep

### PR DESCRIPTION
**Pull request by guswynn**
In preparation for merging https://github.com/MaterializeInc/materialize/pull/18364, and the fact that we are going to being adding `rocksdb` to `mz_storage` soon. This pr adds `rocksdb` as a `dev-dependency` for now (so only `--all-targets` will build it), and adds documentation on what package needs to be installed on linux, which will be announced in `#eng-announce` as this merges

We also need to allow a new duplicate (`bindgen`), until `mysql_common` has a new release. I asked in the rust-mysql glitter chat about when this will happen.

### Motivation

   * This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

**Review by philip-stoev — APPROVED** [review]

File: doc/developer/guide.md
Comment (benesch) [inline doc/developer/guide.md:20]: I think we should rewrite this whole section like so:

```
Materialize depends on several components that are written in C and C++, so
you'll need a working C and C++ toolchain. You'll also need to install:
  * The [CMake] build system
  * libclang
  * PostgreSQL

On macOS, if you install [Homebrew], you'll be guided through the process of
installing Apple's developer tools, which includes a C compiler and libclang.
Then it's a cinch to install CMake and PostgreSQL:
```

File: ci/builder/Dockerfile
Comment (benesch) [inline ci/builder/Dockerfile:155]: Does `--locked` do anything here? There's no Cargo.lock that's pinning the transitive deps, and the top-level version is already pinned.

File: deny.toml
Comment (benesch) [inline deny.toml:25]: Man, rust_mysql_common doesn't even _use_ this dep at build time; it's just for the maintainer. Annoying!

**Review by benesch — APPROVED** [review]
LGTM modulo a slight rewording of the developer guide.

File: deny.toml
Comment (guswynn) [reply deny.toml:25]: :(

**Comment by benesch** [comment]
As you're merging this, can you make sure to post in #eng-announce to warn people that they need to `apt install libclang-dev` if they're compiling Materialize on Linux?

File: ci/builder/Dockerfile
Comment (guswynn) [reply ci/builder/Dockerfile:155]: Actually it does do something, weirdly; this failed to compile because a newer version of one of its deps was broken; I think the `--version` with the `--locked` uses the `Cargo.lock` that existed at that version, or something

Comment (guswynn) [reply ci/builder/Dockerfile:155]: I think so!

**Comment by guswynn** [comment]
@benesch Yes I will annouce!

**Pull request merged**
